### DISCUSSION TITLE
CSV Export -> CSV Import for M2O relationship failing

### DISCUSSION THREAD

**Discussion by chantlong** [directus/directus#14516] [Archive]
Importing a CSV that is exported by Directus to a new table (collection) in Directus fails when I have a M2O relationship.

### To Reproduce

Created a Table with the following data model

TABLE_A
```
id: UUID
parent: M2O (self referencing same collection id of TABLE_A)
type: string
value: string
```
[test 20220718-232645.csv](https://github.com/directus/directus/files/9136251/test.20220718-232645.csv)
I put in some dummy data and exported via Directus.

Create a TABLE_B

```
id: UUID,
type: string,
value: string
parent: M2O (self referencing same collection id of TABLE_B)
```

Import the CSV file created from TABLE_A earlier and click on Start Import and an error toast pops up.

### Errors Shown

Error message is "The file's data structure does not match the collection"

**Comment by rijkvanzanten** [comment]
I'm pretty sure it's because of the import order. The CSV will be parsed and imported sequentially, and in your data dump, the first record has a foreign key to an item that doesn't exist (parent = `653579ca-17c3-4a70-9a67-eb897f4e681b`). If you update the dump (or change the sort order before exporting) to 

```
"id","parent","type","value"
"653579ca-17c3-4a70-9a67-eb897f4e681b",,"food","fruit"
"05690034-042e-4988-b0b4-558bbd83b839","653579ca-17c3-4a70-9a67-eb897f4e681b","fruit_child","grape"
"4068b2bf-09e1-4030-a806-75cdf2614cf0","653579ca-17c3-4a70-9a67-eb897f4e681b","fruit_child","orange"
"1c94d538-b77a-4e7d-ae9f-eda7b65d53f3","05690034-042e-4988-b0b4-558bbd83b839","fruit_in","seed"
```

it imports as expected 👍🏻

**Reply by chantlong** [reply]
I should have noticed that. Thanks!!

**Reply by chantlong** [reply]
Let's say for example the 4th record `fruit_in` is on the 2nd row, it will break because I did not import `fruit_child` first, so I can't just sort by ascending where `parent` is `null` when I have infinitely nested children.

The workaround I can think of is 2 ways.

1. First have Table B `parent` be a regular `input of uuid` just for the import, but I'm not sure how I can later convert back the `parent` to a relation.
2. Export `food` ONLY, `fruit_child` ONLY, `fruit_in` ONLY separately as 3 files, and import them in order.
3. **CURRENT SOLUTION** export the CSV and sort by `type` based on my Custom List (Excel Functionality)

Is there any other solution you can think of?

**Reply by rijkvanzanten** [reply]
Imports use an `upsert`, so there's also an option 4: Import everything without the `parent` field first (create), then import _again_ with the `parent` field to update the now existing items with the parent field (which now is a valid reference)

**Reply by chantlong** [reply]
Ahh that makes sense. Thank you.
